[All Platforms] Rework Window Handle

time-killer-games opened the pull request:
This gives extensions (i.e. *.dll, *.dylib, *.so, etc.) and shell scripts (i.e. execute_shell_for_output, execute_shell, execute_program, etc.) access to window handle on all platforms except Android. Android threw errors for unknown types being used for the android window and surface, which is odd because I'm using the exact types specified by SDL docs; i guess i need to include a proper header but I don't know what that is and I've already tried the header I thought i needed but it didn't work out. so anyway, I commented out android's window handle code as a placeholder file for whenever we do add it correctly. As for the other platforms, this pr also makes the window handle more universal, expecting to use the same variable names for the window handle, and related platform-specific things such as an HINSTANCE for Windows (enigma::hInstance), and a X Window System Display * for Linux (engima::x11::disp), and these variable names are shared across the original platforms as well as their SDL equivalents. So, please do not change these variable namess one place, without changing them everywhere else they are used in the engine.

A follow up pr will need to implement external_call/define/free() on all non-Windows platforms, as well as fix the currently broken Windows implementation; this pr is in preparation for that. Also this pull request statically links SDL on all platforms.

time-killer-games commented:
@JoshDreamland why is it not finding that header? I have all the depends installing in travis properly... or at least I thought I did. Not sure what's wrong with it.

time-killer-games commented:
@JoshDreamland please review kthx

time-killer-games commented:
@JoshDreamland u ween this mudge?

time-killer-games commented:
@RobertBColton Josh wants to know if you are ok with relicensing your libpng code you wrote under whatever SDL uses I think it's MIT or zlib. I want to translate this pr into a contribution to the SDL project itself, because they currently do not have a way to set an x11 icon for Linux games, and I am using both my and your code to do this, but mostly yours.

time-killer-games commented:
@JoshDreamland ready for review, your royal heinieness

JoshDreamland commented:
I thought you were going to remove the SDL integrations from Xlib?

time-killer-games commented:
> I thought you were going to remove the SDL integrations from Xlib?

I kinda did. What you're looking at is widget related stuff, which i thought you would be ok with, because you had previously merged X11 code used in my widgets for checking the KWIN_RUNNING Atom, which would be used regardless of whether the user chose X11 or SDL.

JoshDreamland commented:
I was mostly just confused about what's going on. The flow of code around `window_id_init` is difficult to follow, largely because it introduces a new constraint that some platforms (X11, SDL) must adhere to, but you didn't document what those constraints are, or what platforms they actually apply to.

In general, we avoid creating these couplings between systems—it leads to [spaghetti code](https://en.wikipedia.org/wiki/Spaghetti_code). Particularly with one or more of these systems missing from CI coverage, it's likely to cause breakages, as people won't realize that renaming a method in one system will break another. In general, healthy code means contributors can be confident that changes they make will not introduce unintended side-effects. So the less intimacy you have between systems, the better.

There are, of course, places in the engine where systems require functionality from one another. When we need **all** windowing systems to support a particular routine, we add it to [platforms_mandatory](https://github.com/enigma-dev/enigma-dev/blob/master/ENIGMAsystem/SHELL/Platforms/platforms_mandatory.h). There are functions in that header that are similar to what you seem to want. For instance, [`initGameWindow`](https://github.com/enigma-dev/enigma-dev/tree/master/ENIGMAsystem/SHELL/Platforms/SDL/Window.cpp#L42) is probably the place where you would have SDL stash its information somewhere. Between `initGameWindow`, `initInput`, and `EnableDrawing`, I suspect whatever opportunity you need to to take is covered by one of those methods.

There are other places in the engine where you'll see this kind of intimacy. The biggest are in the "Bridges" and "General" folders. Bridges depend directly on two systems, and access members between each directly to tie things together. Bridges are where you'll find code for binding libraries, like WGL and GLX, which depend on the HWND/X resource and the GL context.

We don't currently have a bridge structure to bind widget systems to window systems ("Platforms"). I am certainly not opposed to adding that bridge system, but we should do so in a way that is analogous to existing bridges, rather than tying code together between two systems.

time-killer-games commented:
@JoshDreamland so then, heinie, what might you suggest I do? (Also, do you mind if I call you heinie for now on?)

RobertBColton commented:
Yeah you can relicense my code but not remove my name and send the fixes upstream to SDL where they belong.

time-killer-games commented:
> Yeah you can relicense my code, but not remove my name, and send the fixes upstream to SDL where they belong.

I wasn't planning on removing your name, this is your code mostly and that would be a rather big asshole move on my part if I did that. Thank you for the permission, that is great news!

JoshDreamland commented:
Is there a reason you can't move that code into the existing `initGameWindow` routine? Or else, invoke that routine from inside that one (so the naming convention is contained within SDL).

time-killer-games commented:
Sorry, that's fine, I didn't understand what you were saying in your initial post about that. I must've read it when I was tired or something.

time-killer-games commented:
@JoshDreamland No, I can't just stash platform specific code in initGameWindow, it relies on X11. I set it up the way it is so that the Makefiles would ensure all that Linux-specific code doesn't attempt to be compiled anywhere but on Linux. If I put that stuff in initGameWindow, Windows and Mac wouldn't build anymore, unless I did platform macros in the code itself and not the Makefiles which I thought you guys were completely against doing.

JoshDreamland commented:
As I said, feel free to define a helper method inside of SDL, call it from initGameWindow, and implement it differently between SDL/Win32 and SDL/X11. The distinction is that other window systems (like Xlib) don't have to worry about constraints within SDL.

time-killer-games commented:
Josh, you aren't making sense here. I already told you how that is not even possible. A helper function won't magically allow X11 code to build ok on Windows and Mac.

time-killer-games commented:
@JoshDreamland ready for review

JoshDreamland commented on ENIGMAsystem/SHELL/Audio_Systems/DirectSound/DSsystem.cpp:
```diff
@@ -70,7 +70,7 @@ void eos_callback(void* soundID, unsigned src) {
 
 int audiosystem_initialize() {
-  if (get_window_handle() == NULL) {
+  if (enigma::hWnd == NULL) {
```
Otherwise, in addition to the breakages, we end up with this line. This is the kind of inter-system dependency I was talking about, earlier. Using `window_handle()` was *somewhat* questionable, but part of ENIGMA's API. Using `enigma::hWnd` is not portable. For example, this system *would* have worked with SDL before. Now it can't, unless SDL *happens* to store a variable by the same name. If that name is changed, this relationship will be broken. Note that SDL is an example. I am not asking you to make it so that using enigma::hWnd works here when building against SDL, but instead to use library methods so arbitrary components will play nicely together.

time-killer-games replied:
SDL does story a variable of the same name because in this pr I added it. There's nothing breaking here. I've already tested it on Linux. If you feel more comfortable with it I can test on Windows too before we hit merge but not only did the appveyor CI pass, I have looked over this code a crap ton of times now to make sure everything is correct based.

JoshDreamland commented on ENIGMAsystem/SHELL/Bridges/Win32-Direct3D11/graphics_bridge.cpp:
```diff
@@ -146,7 +145,6 @@ void WindowResized() {
 }
 
 void EnableDrawing(void* handle) {
-  get_window_handle();
```
You should always be skeptical of deleting a line that you think doesn't do anything. Usually, that's a code smell. In this case, the comment you modified below is clear about this: "This function is called by d3d and dsound to ensure the window handle is set"

If this were done properly, the author would have made a new routine (in Mandatory or in the Win32 bridge header) to make sure that the window was initialized and ready to go. Instead, we ended up with this. But deleting this line will break this system.

time-killer-games replied:
This line in fact doesn't do anything after the changes I made. get_window_handle() has been renamed to window_id_init(); the two functions do the same exact thing. Only now instead of it being called from direct sound and d3d it is being called before both of those things initialized, under initGameWindow() instead, so instead of calling it in multiple irrelevent places, we are calling it once in a relevant place.

time-killer-games commented on ENIGMAsystem/SHELL/Audio_Systems/DirectSound/DSsystem.cpp:
```diff
@@ -17,7 +17,7 @@
 #include "Widget_Systems/widgets_mandatory.h"  // DEBUG_MESSAGE
 
-#include "Bridges/Win32/WINDOWShandle.h" //get_window_handle()
+#include "Platforms/Win32/WINDOWSmain.h" // enigma::hWnd
```
I said that because I didn't know at the time we had anything for this under bridges. I remembered vaguely the pr which did that never got merged, but apparently it did.

time-killer-games commented:
@JoshDreamland before this gets merged and you accuse me of breaking SDL. SDL has been unusable for me for a long time now, dating farther back than any of my recent pr's, including the windowing related ones obviously because I didn't touch SDL's code in those pr's at all. On Windows, the SDL window is always a blank black window for me. On Linux, I get a window client area that looks like the grey fizz on an old TV screen. Just wanted to point out these issues before this gets merged and you blame me for something I didnt break. I just never cared to mention these bugs with SDL until it became a concern that I might get blamed for it when it wasn't me who caused it. Before this pr, I haven't touched SDL hardly at all other than for implemented joysticks half way. No one blame me.

time-killer-games commented:
@RobertBColton it turns out me and Josh were wrong, SDL does in fact have icon support for linux already. Does this mean we should set the license back to GPL? I was going to create a GameMaker extension which uses the code and I shared it on the GameMaker Community in my code snippet topic - https://forum.yoyogames.com/index.php?threads/code-bank-share-useful-code.60575/#post-406540 - Please let me know if you want me to take this down, since the original reason we relicensed it under zlib has been nullified.

RobertBColton commented:
It's ok, I don't mind.

time-killer-games commented:
> It's ok, I don't mind.

Thank you!

fundies commented on ENIGMAsystem/SHELL/Platforms/xlib/Makefile:
```diff
@@ -1,5 +1,5 @@
-override CXXFLAGS +=  -I/usr/X11R6/include -I/usr/include -I/usr/include/libpng
-override CFLAGS += -I/usr/X11R6/include -I/usr/include -I/usr/include/libpng
+override CXXFLAGS +=  -I/usr/X11R6/include -I/usr/include
+override CFLAGS += -I/usr/X11R6/include -I/usr/include
```
cant u use pkgcong here?

time-killer-games replied:
Do you mean I just do:
`override CXXFLAGS += $(pkg-config x11 --cflags --libs)
override CFLAGS += $(pkg-config x11 --cflags --libs)`

Just wanted to make sure I wrote this correctly before I commit.

time-killer-games replied:
That works. Will do.

time-killer-games replied:
@fundies ok. Change made and it's passing. Let me know if you run into any other objections and I'll be happy to fix it accordingly.

time-killer-games commented:
@RobertBColton Josh doesn't want to merge this until people other than myself have tested it on all platforms. But especially Win32. I was wondering if you'd be willing to test it on Win32 and SDL on Windows and check for regressions because apparently Josh doesn't trust me and thinks I'm lying to him when i told him there are none that I could find.

RobertBColton commented:
No, there's a CI testing harness for this purpose, just automate the test.

time-killer-games commented:
@RobertBColton you don't understand, this makes a lot of changes other than just adding icon support. I also moved around a lot of the window handle code to be cross-platform and more efficient, and Josh is worried I broke Direct Sound and  Direct3D for reasons I know are invalid because I took his concerns into consideration when I made these changes but he doesn't believe me. So what he concluded is unless like 10 people test this manually he ain't merging it. If you don't want to or don't have the time, I can't fault you for that. It's ok.

RobertBColton commented:
I'm sorry, but you will continue to elicit that reaction from Josh until you stop conflating unrelated changes together. It's the contributor's responsibility to make their changes digestible for the reviewer.

time-killer-games commented:
@RobertBColton and it's Josh's responsibility to not react that way because it's extremely childish, and your support for him doing this is mind blowing.

RobertBColton commented:
I'm sorry, but I was just giving thoughtful advice on how to get Josh to merge the changes you want. It's the same advice I've given countless times. If it's the case that you didn't want my honest advice, please do not notify me with the at mention again in the future. Thanks!

time-killer-games commented:
@RobertBColton This pr was made a full month ago. I can apply your advice to my future pr's. I can't do that with an existing one. I didn't ask for any advice anyway. I just nicely asked whether you could test it and instead of telling me yes or no you bitched at me over things unrelated to what I was asking.

time-killer-games commented:
@JoshDreamland another two months of this pr has been sitting around with no review. I know you do not like the unrelated issues being solved in one mutual pr. I do not intend to do that anymore going forward, but if you made time to review this, a lot of people would appreciate it. I don't know whether Linux developers care about these things, but I know no sane Windows or Mac developer would even consider publishing a single game or app without a custom icon and would strongly prefer to not use the OS default, which makes a rather poor reflection on the developer, gives a lazy vibe. The other things this pr fixes and adds to are rather important as well. I hope we get this sorted out soon.

JoshDreamland commented:
It's not about how many issues are fixed, Sam; fixing issues is fine. It's about the certifiable correctness of a change. I can certify that your change to X11's makefile is correct; it's a direct improvement, as it removes assumptions we make about libx11 and its dependencies. While I am not well-poised to verify the changes you made to DirectSound, your argument there is compelling; the removal of the dependency on window_handle() being called is an improvement, if that was actually decoupled.

Unfortunately, now DirectSound depends directly on enigma::hWnd, which could change in the future. The Bridges/ folder is the place we define things we consider to be part of the API of the system—notice that DirectSound was including Bridges/Win32/WINDOWShandle.h previously, but now it's including WINDOWSmain directly. This interdependency makes it difficult to make changes to systems in the future, and it's needless in this case because window_handle was already part of our API (and a required part of the EDL API).

Additionally, this is the third PR you have open with a copy of bitmap processing logic in it. How many copies of bitmap processing logic do we need in the engine?

time-killer-games commented:
@JoshDreamland thanks for the explanations. I'll have a look at this and will start working on it probably tmr. For now I want to focus on what we were discussing on hangouts because its code I'm more familiar with (because I've worked on it more recently)

time-killer-games commented:
@JoshDreamland check the messages I sent you on hangouts for more details on why I requested another review without addressing everything you mentioned.

time-killer-games commented:
@JoshDreamland I have addressed all of your concerns in this pr. I have moved the bitmap color conversion logic to image_formats.cpp, I have reverted to the use of the compatibility switcher for window_handle(), and we are no longer including WINDOWSmain.h in DirectSound, as I have reverted it to WINDOWShandle.h. Any other reasons remaining why you should critique this, or can we finally merge this bad boy?

time-killer-games commented:
@JoshDreamland would it help at all if I added the window_handle() changes to a different pr? Unfortunately, that's what the bulk of this pr is actually doing.

time-killer-games commented:
@JoshDreamland before this gets merged im going to compare a test project with this branch and master to make sure this doesn't break OpenGL on Windows. On windows I remember testing this and all versions of OpenGL and and OpenGL ES weren't working on Windows. If I can verify this is also a problem in master than I can know this pr isn't the cause of it. I'll let you know what happens when I have the updated info.

time-killer-games commented:
@JoshDreamland Ok, turns out after testing this again I must've remembered wrong. OpenGL is working just fine in this pr. It's ready for review.

JoshDreamland commented on ENIGMAsystem/SHELL/Platforms/xlib/XLIBwindow.cpp:
```diff
@@ -143,8 +144,39 @@ void destroyWindow() {
+// returns an identifier for the gay window
+// this string can be used in shell scripts
+string window_identifier() {
+  return remove_trailing_zeros((int)enigma::x11::win);
```
Why are there two copies of each of these?

JoshDreamland commented on ENIGMAsystem/SHELL/Platforms/SDL/Cocoa/CocoaWindow.m:
```diff
@@ -0,0 +1,36 @@
+  return [(NSWindow *)cocoa_window_handle() windowNubmer];
+}
+
+long cocoa_window_get_identifier(void *hwnd) {
```
...except here. What happens if a user treats this as a string, and then tries to build their game for Mac? This is one of the reasons I generally don't condone building extensions onto a system that doesn't compile.

time-killer-games replied:
a windowNumber is a long...

time-killer-games replied:
I've created an extension which does this very thing.

JoshDreamland commented on ENIGMAsystem/SHELL/Audio_Systems/DirectSound/DSsystem.cpp:
```diff
@@ -70,7 +70,7 @@ void eos_callback(void* soundID, unsigned src) {
 
 int audiosystem_initialize() {
-  if (get_window_handle() == NULL) {
+  if (enigma::hWnd == NULL || enigma::hInstance == NULL) {
```
Why favor direct use of hWnd to an API method?

time-killer-games replied:
Because Robert said window_handle() is meant to be used by extensions and shouldn't be used by the engine at all.. idk you'd have to ask him I'm just going by what he said in a code comment.

time-killer-games replied:
The real question is, why not?

time-killer-games replied:
@JoshDreamland why wrap those variables in an API function that's just going to hold the same information?

time-killer-games commented on ENIGMAsystem/SHELL/Platforms/SDL/xlib/Window.cpp:
```diff
@@ -0,0 +1,81 @@
+// returns an identifier for the sdl window
+// this string can be used in shell scripts
+string window_identifier() {
+  return remove_trailing_zeros((int)enigma::x11::win);
```
Yeah about the decimal, I realize that now. As for the int, it is an XLIB Window which is just a typedef of an int

time-killer-games commented on ENIGMAsystem/SHELL/Platforms/SDL/Win32/Window.cpp:
```diff
@@ -0,0 +1,78 @@
+// returns an identifier for the sdl window
+// this string can be used in shell scripts
+string window_identifier() {
+  return remove_trailing_zeros((long long)enigma::hWnd);
```
i'm going to address this

time-killer-games commented on ENIGMAsystem/SHELL/Platforms/SDL/Android/Window.cpp:
```diff
@@ -0,0 +1,57 @@
+  SDL_SysWMinfo systemInfo;
+  SDL_VERSION(&systemInfo.version);
+  SDL_GetWindowWMInfo(windowHandle, &systemInfo);
+  ANWin = systemInfo.info.android.window;
```
it fails to build due to unknown functions and types, even though i was going by the SDL docs to do this

time-killer-games commented:
@JoshDreamland why should i use window_handle() instead of `enigma::hWnd`, `enigma::NSWin`, and `enigma::x11::win`. That doesn't exactly solve the platform interdependency problem when we have other info we need to extract from the window other than simply the window handle itself, information such as `enigma::hInstance`, `enigma::WinNum`, and `enigma::x11::disp`...
these things will need to be accessible whether using SDL or otherwise in a lot of cases
you're failing to realize this doesn't solve virtually anything what you are asking me to do

JoshDreamland commented:
In general, if you can get additional type safety from using implementation-specific data, then go for it. But in this case, we're just converting to a boolean, so it doesn't really matter whether it was the original HWND/atom pointer, a `void*`, or a `long`. The test is just `!= 0` (implicitly). So I'd just leave it the way it was (boolean conversion of `window_handle()`), as this is just one less file dependent on internals of other systems. Granted, I don't expect a whole lot of changes to the hWnd/hInstance variables, but this would be just one less place to update them if I'm wrong.

Essentially, it's one less change now, and then one less change in the future if anyone goes to refactor how handles are stored.

time-killer-games commented:
@JoshDreamland I see. So are you implying where it is being tested as a boolean is the only place it should be replaced? If that's the case, I totally misunderstood you. I thought you wanted it to be casted hastily a over the engine

fundies commented on ENIGMAsystem/SHELL/Platforms/SDL/Makefile:
```diff
@@ -7,12 +7,27 @@ override CXXFLAGS += $(shell $(PKG-CONFIG) --cflags sdl2) -DENIGMA_PLATFORM_SDL
 
 ifneq ($(TARGET-PLATFORM), Android)
```
instead of putting these if else chains everywhere we could use more consistent file naming and do something like SOURCES+=  /path/$(PLATFORM)/file.cpp

JoshDreamland replied:
The problem here is that SDL is trying to be its own damned operating system, which normally really helps, but ENIGMA will always offer a rough superset of SDL functionality (ie, a superset except for some known missing features available in various other platforms).

That said... this logic reads strangely, yes.

1. This if statement is checking platform != android instead of checking ==. For ease of skimming, it's usually better to use the positive form (ie, remove the not and swap the two bodies).
1. If the system is Android, we include Android/Window.cpp. Otherwise, we include General/....Posix...shell.cpp? What is going on?
1. The general structure of this code is `{ Android }` `{ Windows }` `{ Not Windows { OS X } { Linux } { FreeBSD } }`.
   a. Why is Android not with the other Not-Windows OSes?
   b. Why is only Android including POSIX/shell.cpp? What causes the other POSIX systems to include it?

time-killer-games replied:
Are you sure $(PLATFORM) will ever equal xlib instead of Linux or FreeBSD?

fundies replied:
that was just an example and not meant to be taken verbatim. my point was you could avoid the if chain with proper folder names and variables. So if we needed threads.cpp from each platform for sdl you could use SOURCES += $(plat)/threads.cpp so for win32 it would look in win32/threads.cpp and for xlib it would grab xlib/threads.cpp

fundies commented on ENIGMAsystem/SHELL/Platforms/SDL/Win32/Window.cpp:
```diff
@@ -0,0 +1,69 @@
+
+#if GM_COMPATIBILITY_VERSION <= 81
+unsigned long long window_handle() {
+  return (unsigned long long)enigma::hWnd;
```
try to use static_casts as they usually result in compiler errors rather than crashes

JoshDreamland replied:
In general, it's a good rule; Google insists on it. Personally, for integer casts, I don't care.

fundies commented on ENIGMAsystem/SHELL/Platforms/SDL/Window.h:
```diff
@@ -14,6 +14,7 @@ namespace enigma
     extern std::unordered_map<SDL_Keycode,int> keymap;
     extern std::unordered_map<SDL_Keycode, int> inverse_keymap;
   }
+  void window_id_init();
```
is this sdl specific? I feel like it should be declared in general

JoshDreamland replied:
I believe it's specific to subsystems of SDL. Other systems only have one way of implementing their window handle logic, because it happens naturally while creating the window, but SDL needs to pretend to be each other operating system itself.

time-killer-games replied:
It's declaired on the other platforms a different way iirc. I did this for SDL so that window_handle could be captured differently per platform

fundies commented on ENIGMAsystem/SHELL/Platforms/SDL/xlib/Window.cpp:
```diff
@@ -0,0 +1,72 @@
+
+#if GM_COMPATIBILITY_VERSION <= 81
+unsigned long long window_handle() {
+  return (unsigned long long)enigma::x11::win;
```
c++ cast

fundies commented on ENIGMAsystem/SHELL/Platforms/SDL/xlib/Window.cpp:
```diff
@@ -0,0 +1,72 @@
+// returns an identifier for the sdl window
+// this string can be used in shell scripts
+string window_identifier() {
+  return std::to_string((unsigned long)enigma::x11::win);
```
c++ cast

fundies commented on ENIGMAsystem/SHELL/Platforms/SDL/xlib/Window.cpp:
```diff
@@ -0,0 +1,72 @@
+
+// returns an identifier for certain window
+// this string can be used in shell scripts
+string window_get_identifier(void *hwnd) {
```
consider using a typedef or a using here instead of void

JoshDreamland replied:
He's trying to make this an ENIGMA API call, afaict. A typedef should still work fine, and will help it look better in the IDE, though, so why not.

fundies commented on ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSmain.cpp:
```diff
@@ -49,14 +51,16 @@ namespace enigma_user {
 HDC window_hDC;
 HANDLE mainthread;
 
+LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam);
```
This line might be incorrect for msvc. I vaguely remember fixing it in my cmake pr.

time-killer-games replied:
What about it is wrong? Looking here it seems correct; we just used a different function name for the proc https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms633573(v%3Dvs.85)

fundies replied:
I was wrong it was a similar line tho https://github.com/enigma-dev/enigma-dev/pull/1490/files#diff-5de129638eaf07e20162090473aa3bc3R65 .

fundies commented on ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSwindow.cpp:
```diff
@@ -81,11 +82,23 @@ unsigned long long window_handle() {
-void* window_handle() {
-  return enigma::hWnd;
+void *window_handle() {
+  return (void *)enigma::hWnd;
```
again id like to see a typedef here and a c++ cast

fundies commented on ENIGMAsystem/SHELL/Widget_Systems/xlib/Info/About.ey:
```diff
@@ -1,15 +1,16 @@
 ---
 
-Name: Linux X11 Widgets
+Name: Linux/FreeBSD X11 Widgets
```
no need to list OS in name as osx and likely other oses can run xlib stuff too

time-killer-games replied:
Yeah I'll make it just say "X11" and "X11 Widgets" for the related Platform and Widget System

time-killer-games replied:
I actually think I meant to do this change in my BSD pr

fundies commented on ENIGMAsystem/SHELL/Platforms/SDL/Android/Window.cpp:
```diff
@@ -0,0 +1,57 @@
+// called from initGameWindow()
+// capture sdl window hwnd/surf
+void window_id_init() {
+  /*
```
youshould mark stuff like this with TODO or FIXME so it's easier for people to find

JoshDreamland commented on ENIGMAsystem/SHELL/Widget_Systems/xlib/kdialog.cpp:
```diff
@@ -132,7 +132,7 @@ static int show_message_helperfunc(string message) {
 
   str_command = string("kdialog ") +
-  string("--attach=") + std::to_string((int)enigma::x11::win) + string(" ") + str_cancel +
+  string("--attach=") + std::to_string((unsigned long)enigma::x11::win) + string(" ") + str_cancel +
```
This is probably a good place to use your new `window_identifier` call instead of directly accessing window primitives.

time-killer-games replied:
yee, will do there, son!

time-killer-games commented:
@fundies how do I fix these failing checks? I'm not familiar with this new CI system we got going on here.

time-killer-games commented:
@JoshDreamland fundies said reinterpret_cast and C-style casts are dangerous so what alternative do we have to cast unsigned long long to/from pointers when static_cast throws compile errors?

time-killer-games commented:
@JoshDreamland this should be ready to be merged unless you have any other objections.

time-killer-games commented:
@JoshDreamland ready for merging

RobertBColton commented:
Was the compatibility macro factored into this pull request like #1471 had?

time-killer-games commented:
@RobertBColton if i'm understanding your question correctly, yes.
